### DISCUSSION TITLE
useQueries combined data is never undefined in contrast to useQuery

### DISCUSSION THREAD

**Discussion by nop33** [TanStack/query#7987] [Q&A]
I run into a type error when one of my `combine` functions of a `useQueries` is returning an object (by reducing the results array into a object so that I can use IDs as keys for faster search) and I try to access a property of that object. It happened that that object was `undefined`. So, I am curious, shouldn't `data` returned by `combine` also be `undefined`, like the `data` from `useQuery` is?

Here is a simple example with `number`:

[Typescript playground](https://www.typescriptlang.org/play/?#code/LAKAlgtgDg9gTgFwAQG8kFcDOBTAiu7OMbTAGgx30IE8kBfJAMzhgiQHIABBAQwDtMvAMYBrAPRxsPIQgC0ARwJxq7ANyhQ2AB6xESITAHIs2APJ9sAFQAWYPgHMkAXiQAKAJTOAfKlBJ-+oaCSHzobC4AjH4BBkaoSAAmPLz0zhR4StSuKNEBAYo0ANLY1ABcSADa7Ai2Dpjs5KEQALqkuXlIBcoAYnzlHt5IAAosEGA4AHSSmDAANgBu2K5N7u10qyDtYmKJyTxI40gARE0ARoRIAD4YfAnYjHbYCUdIM0gAkvr81UgAVljIIQ8WazJgwGCnHhwVxJXieADuYBqMHQCC2OyE1mwojsjjAjF2KUOdweFgS5HhtkxB0wSG0UGxCCe5GBoPs4ISE3a+LcsJ4nhyIA6AUY4Mh0L5GzydHaoBlmxA9PggKCxhwAFl+NQbLjMAAlJ7oIRPNIDJw+QV5WLBJq0lwVCLkABM5AAzM12tbkGg+akXCYqEQSNl2nkusRMOVbRMIDwoK5lmFPOa3JbhR0utRimVKtVavZ6o0wq1Q+nM71+smfCNWONsFMSHNFomIFL0+t3G0hemDBBTo9+tN0LMEJgq6nS8K+eUhyPMA2Ekalq5pEJyLOEOPV0gANRuDcTX0AfiPSAADJ3z23hes1ht0YT9ocLIs4Md0Ld7o9nuRTqiDnwSCSMCSK0EiSCIiCSDnAcBI1NgnRKBGSDWDwiwPnwMDINQ2DIJICDoHAZJIPwCSBH2jwoTwmB8D8cAfuQbw1IcsyUZBoI1Cw8IkUgCDUAyD6ECwb52LxqGAtRCGUtggHaNiqK4u0ooQlCMJ7Bs8qgF6YIqW+LiuPMwIEFGYTnHA45ptacz1rMMD2AZRnYBpQA)

```ts

import { useQueries, useQuery } from '@tanstack/react-query';

export const useOneThing = () => {
    const num = 1
    const { data } = useQuery({
        queryKey: ['things', num],
        queryFn: () => Promise.resolve(num)
    })

    // data is "number | undefined" so I can't just call foobar(data) without
    // checking if data is defined, which is expected, all good.
    if (data) {
        foobar(data)
    }
    
}

export const useManyThingsReduced = () => {
    const nums = [1, 2, 3]
    const { data } = useQueries({
        queries: nums.map((num) => ({
            queryKey: ['things', num],
            queryFn: () => Promise.resolve(num)
        })),
        combine: (results) => ({
            data: results.reduce((acc, result) => acc + (result.data ?? 0), 0)
        })
    })

    // data is never "undefined", but in reality it will be if the queries have
    // not yet returned and combine hasn't run, so this line will throw a type
    // error in that case when executing
    foobar(data)
}

const foobar = (value: number) => {
  console.log(value)
}
```

**Comment by TkDodo** [comment]
yes, this is likely wrong. Would you like to fix it & add some tests?

**Reply by TkDodo** [reply]
on second thought - can you make a runtime reproduction for this please? We always run `combine`, even on the first render cycle. You would likely get `results` into combine where `data` is still `undefined`, but that means you should have an array with `data: 0` and not `undefined` as a result of `useQueries` ....

**Reply by nop33** [reply]
That's indeed what I get (`0`) when I try to reproduce the type error I got in my application. I will keep trying to reproduce, maybe I'll share a more realistic example taken directly from my app. Thanks for your response.

**Reply by nop33** [reply]
I really can't reproduce it. Here's what I tried: https://stackblitz.com/edit/vitejs-vite-31qcuh?file=src%2FApp.tsx

I'll keep an eye on it and come back to this thread if I see something. Thanks a lot, much obliged!
